lib/srv/discovery/fetchers/azure-sync/azure-sync_test.go

```diff
@@ -0,0 +1,227 @@
+	role_name := "test_role_name"
+	action1 := "Microsoft.Compute/virtualMachines/read"
+	action2 := "Microsoft.Compute/virtualMachines/*"
+	action3 := "Microsoft.Compute/*"
+	return &armauthorization.RoleDefinition{
+		ID:   &id,
+		Name: &name,
+		Properties: &armauthorization.RoleDefinitionProperties{
+			Permissions: []*armauthorization.Permission{
+				{
+					Actions: []*string{&action1, &action2},
+				},
+				{
+					Actions: []*string{&action3},
+				},
+			},
+			RoleName: &role_name,
+		},
+	}
+}
+
+func newRoleAssign(id string, name string) *armauthorization.RoleAssignment {
+	scope := "test_scope"
+	principalId := "test_principal_id"
+	roleDefId := "test_role_def_id"
+	return &armauthorization.RoleAssignment{
+		ID:   &id,
+		Name: &name,
+		Properties: &armauthorization.RoleAssignmentProperties{
+			PrincipalID:      &principalId,
+			RoleDefinitionID: &roleDefId,
+			Scope:            &scope,
+		},
+	}
+}
+
+func newVm(id string, name string) *armcompute.VirtualMachine {
+	return &armcompute.VirtualMachine{
+		ID:   &id,
+		Name: &name,
+	}
+}
+
+func TestPoll(t *testing.T) {
+	roleDefs := []*armauthorization.RoleDefinition{
+		newRoleDef("id1", "name1"),
+	}
+	roleAssigns := []*armauthorization.RoleAssignment{
+		newRoleAssign("id1", "name1"),
+	}
+	vms := []*armcompute.VirtualMachine{
+		newVm("id1", "name2"),
+	}
+	roleDefClient := testRoleDefCli{}
+	roleAssignClient := testRoleAssignCli{}
+	vmClient := testVmCli{}
+	fetcher := Fetcher{
+		Config:           Config{SubscriptionID: "1234567890"},
+		lastResult:       &Resources{},
+		roleDefClient:    &roleDefClient,
+		roleAssignClient: &roleAssignClient,
+		vmClient:         &vmClient,
+	}
+	ctx := context.Background()
+	allFeats := Features{
+		RoleDefinitions: true,
+		RoleAssignments: true,
+		VirtualMachines: true,
+	}
+	noVmsFeats := allFeats
+	noVmsFeats.VirtualMachines = false
+
+	tests := []struct {
+		returnErr   bool
+		roleDefs    []*armauthorization.RoleDefinition
+		roleAssigns []*armauthorization.RoleAssignment
+		vms         []*armcompute.VirtualMachine
+		feats       Features
+	}{
+		// Process no results from clients
+		{
+			returnErr:   false,
+			roleDefs:    []*armauthorization.RoleDefinition{},
+			roleAssigns: []*armauthorization.RoleAssignment{},
+			vms:         []*armcompute.VirtualMachine{},
+			feats:       allFeats,
+		},
+		// Process test results from clients
+		{
+			returnErr:   false,
+			roleDefs:    roleDefs,
+			roleAssigns: roleAssigns,
+			vms:         vms,
+			feats:       allFeats,
+		},
+		// Handle errors from clients
+		{
+			returnErr:   true,
+			roleDefs:    roleDefs,
+			roleAssigns: roleAssigns,
+			vms:         vms,
+			feats:       allFeats,
+		},
+		// Handle VM features being disabled
+		{
+			returnErr:   false,
+			roleDefs:    roleDefs,
+			roleAssigns: roleAssigns,
+			vms:         vms,
+			feats:       noVmsFeats,
+		},
+	}
+
+	for _, tt := range tests {
+		// Set the test data
```

Comment:
Usually there's a call to `t.Run("test name", ...`.
This way each record in tests is treated as a unit test and is reported as such (can also be parallelized)